### ISSUE TITLE
Anomaly "Uncaught exception Not_found." from some kind of type error

### ISSUE THREAD

**Issue by elidupree** [coq/coq#13766]
#### Description of the problem

This code (which is not type-correct; the `rew` should be `rew <-`, and if you change it to `rew <-`, it runs successfully):
```coq
Set Implicit Arguments.
Set Asymmetric Patterns.
Require Import Unicode.Utf8.

Import EqNotations.

Inductive NonDeterministicHistory EffectsState (initialState : EffectsState) (step : EffectsState → EffectsState → Prop) : EffectsState → Type :=
| Start : NonDeterministicHistory initialState step initialState
| Step : ∀ old next , step old next → NonDeterministicHistory initialState step old → NonDeterministicHistory initialState step next.

Inductive SubHistory EffectsState SubEffectsState (stateMapping : EffectsState → SubEffectsState) (initialState : EffectsState) (step : EffectsState → EffectsState → Prop) (subStep : SubEffectsState → SubEffectsState → Prop) (latest : EffectsState) : ∀ (history : NonDeterministicHistory initialState step latest) (subInitialState : SubEffectsState), NonDeterministicHistory subInitialState subStep (stateMapping latest) → Type :=
| SubSkip : ∀ old (s : step old latest)
    (prevHistory : NonDeterministicHistory initialState step old)
    (subPrevHistory : NonDeterministicHistory (stateMapping old) subStep (stateMapping old))
    (noChange : stateMapping latest = stateMapping old),
    SubHistory stateMapping prevHistory subPrevHistory →
    SubHistory stateMapping (Step latest s prevHistory) (rew noChange in subPrevHistory).
```
spits out this error when I run it in coqide:
```
Anomaly "Uncaught exception Not_found." Please report at http://coq.inria.fr/bugs/.
```

#### Coq Version

The Coq Proof Assistant, version 8.12.2 (December 2020)
compiled on Dec 26 2020 12:26:31 with OCaml 4.07.1

Running in coqide on Windows 10.

**Comment by olaure01** [comment]
Manually minimized example (tested with same problematic behavior on `Coq 8.13.0`):
```coq
Import EqNotations.

Inductive Bug : 0 <= 1 -> Prop :=
| CBug : forall b (Pb : 0 <= (S b)) (Heq : S b = 1), Bug (rew <- Heq in Pb).
```

**Comment by ppedrot** [comment]
Here is a backtrace:
```
Raised at file "kernel/environ.ml", line 161, characters 29-44
Called from file "engine/eConstr.ml" (inlined), line 807, characters 51-67
Called from file "pretyping/retyping.ml", line 125, characters 34-52
Called from file "pretyping/evarsolve.ml", line 811, characters 8-53
Called from file "pretyping/evarsolve.ml", line 1486, characters 13-61
Called from file "pretyping/evarconv.ml", line 1742, characters 22-69
Called from file "pretyping/evarconv.ml", line 1751, characters 16-21
Called from file "pretyping/evarconv.ml", line 1773, characters 22-78
Called from file "pretyping/evarconv.ml", line 1797, characters 29-49
Called from file "vernac/comInductive.ml", line 165, characters 10-84
...
```
Looks like a rel is fetched inside the wrong environment.

**Comment by SkySkimmer** [comment]
It seems a rel becomes a candidate somehow in second_order_matching.
Then later we try to use it and die.

**Comment by herbelin** [comment]
Both the original and simplified examples are fixed since 8.15 (returning instead an error).
